[GPU Process] (REGRESSION 262607@main): Copy pixels of CGImage only if it is not a sub-image

shallawa opened the pull request:
#### 58702859275c833b40676afa1d2001fae6569959
<pre>
[GPU Process] (REGRESSION 262607@main): Copy pixels of CGImage only if it is not a sub-image
<a href="https://bugs.webkit.org/show_bug.cgi?id=255677">https://bugs.webkit.org/show_bug.cgi?id=255677</a>
rdar://108072383

Reviewed by Simon Fraser.

In ShareableBitmap::createFromImagePixels(), CGImage and its CGDataProvider
may disagree about how many bytes are in the pixels buffers. CGImage is most
likely a sub-image of another CGImage which inherits its meta data but not
the full pixels buffer.

In this case we should fallback to ShareableBitmap::createFromImageDraw().
Ortherwise  ShareableBitmapConfiguration will falsely describe the data in the
SharedMemory.

* Source/WebKit/Shared/cg/ShareableBitmapCG.cpp:
(WebKit::ShareableBitmap::createFromImagePixels):

Canonical link: <a href="https://commits.webkit.org/264021@main">https://commits.webkit.org/264021@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d51e0cd6dcf9383497ece28e619b890691542a37

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6661 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6842 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8032 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6748 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6612 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9640 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6566 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/6615 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5846 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8111 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4079 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/5999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5905 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8217 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6389 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5795 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1521 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9954 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/cg/ShareableBitmapCG.cpp:
```diff
@@ -119,17 +119,22 @@ RefPtr<ShareableBitmap> ShareableBitmap::createFromImagePixels(NativeImage& imag
     if (!bytes || !sizeInBytes)
         return nullptr;
 
+    // CGImage and its CGDataProvider disagree about how many bytesPerRow are in pixles buffer.
```
"pixles"

smfr reviewed (commented):
I think we need to find an image that triggers this so we can write a layout test.

smfr commented on Source/WebKit/Shared/cg/ShareableBitmapCG.cpp:
```diff
@@ -119,17 +120,29 @@ RefPtr<ShareableBitmap> ShareableBitmap::createFromImagePixels(NativeImage& imag
 
     const auto* bytes = reinterpret_cast<const uint8_t*>(CFDataGetBytePtr(pixels.get()));
-    auto sizeInBytes = CFDataGetLength(pixels.get());
+    unsigned sizeInBytes = CFDataGetLength(pixels.get());
```
This may overflow.

shallawa committed:
[GPU Process] (REGRESSION 262607@main): Copy pixels of CGImage only if it is not a sub-image
https://bugs.webkit.org/show_bug.cgi?id=255677
rdar://108072383

Reviewed by Simon Fraser.

In ShareableBitmap::createFromImagePixels(), CGImage and its CGDataProvider
may disagree about how many bytes are in the pixels buffers. CGImage is most
likely a sub-image of another CGImage which inherits its meta data but not
the full pixels buffer.

In this case we should fallback to ShareableBitmap::createFromImageDraw().
Ortherwise  ShareableBitmapConfiguration will falsely describe the data in the
SharedMemory.

* Source/WebKit/Shared/cg/ShareableBitmapCG.cpp:
(WebKit::ShareableBitmap::createFromImagePixels):

Canonical link: https://commits.webkit.org/264021@main
